feat(treesitter): disable captures and patterns on TSQuery

ian-h-chamberlain opened the pull request:
Part 1 of #32729 

These APIs are already exposed by tree-sitter API, so this is just a
simple binding to expose them to Lua. Since they are pretty low-level
and not reversible, leave it mostly undocumented for now.

This also adds the pattern indices to the output of `:Inspect!` so that it's actually feasible to identify a given pattern and disable it.

ian-h-chamberlain committed:
feat(lua): disable_{capture,pattern} on TSQuery

These APIs are already exposed by tree-sitter API, so this is just a
simple binding to expose them to Lua. Since they are pretty low-level
and not reversible, leave it undocumented for now (most users probably should
prefer to use e.g. an unconditionally false predicate for disabling patterns).

ian-h-chamberlain committed:
feat(treesitter): add pattern ID to inspect output

Without this, users will not really have any convenient way to figure
out a pattern's ID to use the new disable_pattern() method.

ian-h-chamberlain committed:
test(treesitter): add tests for query disabling

Test that disable_capture() and disable_pattern() work as expected

clason commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -288,15 +288,19 @@ function M.get_captures_at_pos(bufnr, row, col)
-            matches,
-            { capture = capture, metadata = metadata, lang = tree:lang(), id = id }
-          )
+          local _, pattern_i = match:info()
```
```suggestion
          local _, pattern_id = match:info()
```

ian-h-chamberlain replied:
(replying here for threading purposes)

> (Are the pattern ids stable?)

This is a good question and one I also had when I started implementing this... I haven't really found any specific documentation that indicates "stability" of these but from what I understand:

1. This pattern ID is really just an index, so given a query like `(foo) (bar) (baz)` you get IDs of `0`, `1`, `2`, respectively
	- Because nvim may concatenate queries together (e.g. `;extends`, `;inherits`), any changes in query file load order or similar mechanisms has the potential to change the indices
2. Pattern IDs are used in a number of public tree-sitter APIs (like `TSQueryMatch` and `ts_query_disable_pattern`), so I think they can probably be considered as stable as the rest of the API?

clason commented:
I think the exposed functions _should_ be documented, especially if you link them from `news.txt`. A small example snippet would also go a long way.

clason commented on runtime/doc/news.txt:
```diff
@@ -417,6 +417,10 @@ TREESITTER
   code block fence lines vertically.
 • |vim.treesitter.language.inspect()| shows additional information, including
   parser version for ABI 15 parsers.
+• New |TSQuery:disable_pattern()| and |TSQuery:disable_capture()| to turn off
```
```suggestion
• |TSQuery:disable_pattern()| and |TSQuery:disable_capture()| to turn off
```

clason replied:
Those links won't work without actual documentation...

justinmk replied:
Just noticed that docs validation isn't working. Working on a fix

ian-h-chamberlain replied:
Ah, I thought it might just fall back to highlighting if there weren't actual docs to link to and I only really added a `news.txt` entry since the check failed on my first push.

Currently it appears there is no documentation at all for `TSQuery` (although it's referenced in some docs), so I'm not sure what the best way to document these new methods would be – would I add a new class type in `query.lua`, or break out a new file in `runtime/lua/vim/treesitter/_meta` or something?

I guess this is related also to #26721 (exposing more of the `TS*` classes in documentation)?

ian-h-chamberlain replied:
Pushed an update which adds `TSQuery` docs to the `lua-treesitter-query` section; it didn't seem to make sense to follow exactly the format already used by `TSNode`/`TSTree` because they have their own sections and there is already a (non-generated) `TREESITTER QUERIES` section, but please let me know if you think it deserves its own section or any wording changes.

clason commented on runtime/doc/news.txt:
```diff
@@ -417,6 +417,10 @@ TREESITTER
   parser version for ABI 15 parsers.
+• New |TSQuery:disable_pattern()| and |TSQuery:disable_capture()| to turn off
+  a specific pattern or capture in a query.
+• |vim.treesitter.get_captures_at_pos()| now returns the `pattern_id` of the
```
```suggestion
• |vim.treesitter.get_captures_at_pos()| returns the `pattern_id` of the
```

justinmk commented on test/functional/treesitter/query_spec.lua:
```diff
@@ -911,5 +911,65 @@ void ui_refresh(void)
+      eq({ { id = 1, pattern = 1 }, { id = 2, pattern = 3 } }, result)
+    end)
+
+    it('supports disabling captures', function()
```
this code is 99% identical to the above, please factor out into a local function.

ian-h-chamberlain committed:
docs(treesitter): add TSQuery and Query docs

Query was already annotated but marked @nodoc; TSQuery is now also
documented with the new functionality and its docs appended to the
lua-treesitter-query section.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -427,6 +427,10 @@ TREESITTER
+• |TSQuery:disable_pattern()| and |TSQuery:disable_capture()| to turn off
+  a specific pattern or capture in a query.
+• |vim.treesitter.get_captures_at_pos()| returns the `pattern_id` of the
+  pattern used to match each capture.
```
add to the existing item: https://github.com/neovim/neovim/blob/37786ccfcf07706c7524cdeef321bb4defba45b2/runtime/doc/news.txt#L416

justinmk commented on runtime/doc/treesitter.txt:
```diff
@@ -1514,6 +1536,56 @@ set({lang}, {query_name}, {text})                 *vim.treesitter.query.set()*
+    |vim.treesitter.query.get_files()|).
+
+    Example: To disable `|` links in vimdoc but keep other `@markup.link`s
+    highlighted: >lua
```
Nice!

justinmk reviewed (approved):
Looks great. Will wait for @clason to merge

ian-h-chamberlain commented:
Thanks for the quick reviews and merge!
